server/backup_runner.go

```diff
@@ -70,6 +74,101 @@ func (r *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
+			return
+		}
+		for _, backup := range backupList {
+			backupTime := time.Unix(backup.UpdatedTs, 0)
+			expireTime := backupTime.Add(time.Duration(bs.RetentionPeriodTs))
+			if time.Now().After(expireTime) {
+				// update metadata
+				archive := api.Archived
+				backupPatch := api.BackupPatch{
+					ID:        backup.ID,
+					RowStatus: &archive,
+				}
+				if _, err := r.server.store.PatchBackup(ctx, &backupPatch); err != nil {
+					log.Error("Failed to update status for deleted backup", zap.String("name", backup.Name), zap.String("database", bs.Database.Name))
+					continue // next backup
+				}
+
+				// delete backup file
+				backupFilePath := getBackupAbsFilePath(r.server.profile.DataDir, bs.DatabaseID, backup.Name)
+				if err := os.Remove(backupFilePath); err != nil {
+					log.Error("Failed to delete an expired backup file.", zap.String("path", backupFilePath), zap.Error(err))
+				}
+				log.Info("Deleted expired backup file.", zap.String("path", backupFilePath))
+			}
+		}
+	}
+
+	log.Debug("Deleting expired MySQL binlog files.")
+	instanceList, err := r.server.store.FindInstance(ctx, &api.InstanceFind{})
+	if err != nil {
+		log.Error("Failed to find non-archived instances.", zap.Error(err))
+		return
+	}
+
+	for _, instance := range instanceList {
+		if instance.Engine != db.MySQL {
+			continue
+		}
+		backupSettingList, err := r.server.store.FindBackupSetting(ctx, api.BackupSettingFind{InstanceID: &instance.ID})
+		if err != nil {
+			log.Error("Failed to find backup settings for instance.", zap.String("instance", instance.Name), zap.Error(err))
+			continue // next instance
+		}
+		maxRetentionPeriodForInstance := math.MaxInt
+		for _, bs := range backupSettingList {
+			if bs.RetentionPeriodTs != api.BackupRetentionPeriodUnset && bs.RetentionPeriodTs < maxRetentionPeriodForInstance {
+				maxRetentionPeriodForInstance = bs.RetentionPeriodTs
+			}
+		}
+		if maxRetentionPeriodForInstance == math.MaxInt {
+			log.Debug("All the databases in instance have unset retention period. Skip deleting binlog files.", zap.String("instance", instance.Name))
+			continue // next instance
+		}
+
+		log.Debug("Deleting old binlog files for MySQL instance.", zap.String("instance", instance.Name))
+		binlogDir := getBinlogAbsDir(r.server.profile.DataDir, instance.ID)
+		binlogFileInfoList, err := ioutil.ReadDir(binlogDir)
+		if err != nil {
+			log.Error("Failed to read backup directory.", zap.String("path", binlogDir), zap.Error(err))
+			continue // next instance
+		}
+		for _, binlogFileInfo := range binlogFileInfoList {
+			if _, err := mysql.GetBinlogNameSeq(binlogFileInfo.Name()); err != nil {
+				log.Warn("Found an irregular file in binlog directory.", zap.String("path", binlogFileInfo.Name()))
+				continue // next binlog file
+			}
+			expireTime := binlogFileInfo.ModTime().Add(time.Duration(maxRetentionPeriodForInstance))
```

Comment:
1. Fixed duration using time.Second.
2. Checked that timezone is not relevant because we are using the local time of bytebase plus a duration, which is also the local time. We should pay attention to the timezone when using cloud storage.
3. There's not a proper way of getting binlog last update time from MySQL server. Added comment about this.
4. Refactored and extracted function for purging binlog files.